docs/org.eclipse.n4js.spec/chapters/12_components/components.adoc

```diff
@@ -512,9 +510,59 @@ Each listed project in the "typeDependencies" section of a project's `package.js
 . For each listed type dependency, a corresponding dependency (in the \[dev\]dependencies section) must be declared, whose "name" matches the "definesPackage" property value of the specified type dependency.
 --
 
+
+
 [.language-n4js]
-== Component Dependencies
+== Support for NPM Scopes
+
+NPM supports a namespace concept for npm packages. Such namespaces are called "scopes". For details see
+https://docs.npmjs.com/misc/scope and https://docs.npmjs.com/getting-started/scoped-packages.
+In N4JS, this is supported too.
+
+Terminology:
+
+1. A project's _plain project name_ is its name without mentioning the project's scope (if any),
+   e.g. `myProject`.
+2. A project's _scope name_ is the name of the npm scope a project resides in, including a leading `@`.
+   E.g. `@myScope`.
+3. A project's _N4JS project name_ is its plain project name, prefixed by its scope name (if any),
+   separated by a `/`. For unscoped projects, this is identical to the plain project name.
+   E.g., `myProject` (if unscoped), `@myScope/myProject` (if scoped).
+4. A project's _Eclipse project name_ is an ancillary name used only within the Eclipse UI for
+   the project in the workspace. It is equal to the N4JS project name, except that `_` instead of `/` is
+   used as separator between the scope and plain project name.
+   E.g., `myProject` (if unscoped), `@myScope_myProject` (if scoped).
```

Comment:
so if _plain project name_ is `my_project` and _scope name_ is `my_scope` The eclipse name is `@my_scope_my_project`. How do IDE know that this has to mapped to `@my_scope/my_project` and not `@my/scope_my_project?

Reply:
valid npm names are "specified" here: https://www.npmjs.com/package/validate-npm-package-name

Reply:
_ should conform to_ sure.
So looking at https://replicate.npmjs.com/_all_docs I see **18535 npm packages have underscore in name**

You can check this by running this test in N4JS:
```typescript
    public async importCatalogTest2() {
        const resp = await fetch("https://replicate.npmjs.com/_all_docs");
        const data = await resp.json();
        const rows = ((data as Object)['rows'] as [Object]);
        const namesWithUnderscore = rows.filter((o) => { const name = o['id'] as string; return name.indexOf('_') != -1; }).map((o)=> o['id']);
        console.log(`${namesWithUnderscore.length} of ${rows.length} npm packages have underscore in name:`);
        namesWithUnderscore.forEach((name, index)=>console.log(" ",index, "::", name))
    }
```
(needs `n4js-runtime-fetch`)

Reply:
also what about underscores in scope names? e.g.
https://www.npmjs.com/package/@(._.)/oooooo

Reply:
Underscore in package names does not hurt, but in scope names it is an issue. I agree, we cannot use underscore as separator. Interestingly, many package/scope names on npm are actually illegal according to their specifications (maybe for historic reasons). We need a character that does not appear in scope names (package names don't matter) AND is URL-safe (for internal Eclipse reasons).

Anyways, thanks for the above snippet for checking npm names. Very helpful.